modules/javafx.graphics/src/main/java/javafx/scene/layout/BackgroundPosition.java

```diff
@@ -175,6 +193,83 @@ public BackgroundPosition(@NamedArg("horizontalSide") Side horizontalSide, @Name
+        Objects.requireNonNull(endValue, "endValue cannot be null");
+
+        if (t <= 0) {
+            return this;
+        }
+
+        if (t >= 1) {
+            return endValue;
+        }
+
+        double newHorizontalPosition, newVerticalPosition;
+        boolean newHorizontalAsPercentage, newVerticalAsPercentage;
+        Side newHorizontalSide, newVerticalSide;
+
+        if (this.horizontalSide == endValue.horizontalSide
+                && this.horizontalAsPercentage == endValue.horizontalAsPercentage) {
+            newHorizontalPosition = Utils.interpolate(this.horizontalPosition, endValue.horizontalPosition, t);
+            newHorizontalAsPercentage = this.horizontalAsPercentage;
+            newHorizontalSide = this.horizontalSide;
+        } else if (t < 0.5) {
+            newHorizontalPosition = this.horizontalPosition;
+            newHorizontalAsPercentage = this.horizontalAsPercentage;
+            newHorizontalSide = this.horizontalSide;
+        } else {
+            newHorizontalPosition = endValue.horizontalPosition;
+            newHorizontalAsPercentage = endValue.horizontalAsPercentage;
+            newHorizontalSide = endValue.horizontalSide;
+        }
+
+        if (this.verticalSide == endValue.verticalSide
+                && this.verticalAsPercentage == endValue.verticalAsPercentage) {
+            newVerticalPosition = Utils.interpolate(this.verticalPosition, endValue.verticalPosition, t);
+            newVerticalAsPercentage = this.verticalAsPercentage;
+            newVerticalSide = this.verticalSide;
+        } else if (t < 0.5) {
+            newVerticalPosition = this.verticalPosition;
+            newVerticalAsPercentage = this.verticalAsPercentage;
+            newVerticalSide = this.verticalSide;
+        } else {
+            newVerticalPosition = endValue.verticalPosition;
+            newVerticalAsPercentage = endValue.verticalAsPercentage;
+            newVerticalSide = endValue.verticalSide;
+        }
+
+        if (isSame(newHorizontalSide, newHorizontalPosition, newHorizontalAsPercentage,
+                   newVerticalSide, newVerticalPosition, newVerticalAsPercentage)) {
+            return this;
+        }
+
+        if (endValue.isSame(newHorizontalSide, newHorizontalPosition, newHorizontalAsPercentage,
+                            newVerticalSide, newVerticalPosition, newVerticalAsPercentage)) {
+            return endValue;
+        }
+
+        return new BackgroundPosition(
+            newHorizontalSide, newHorizontalPosition, newHorizontalAsPercentage,
+            newVerticalSide, newVerticalPosition, newVerticalAsPercentage);
+    }
+
+    private boolean isSame(Side horizontalSide, double horizontalPosition, boolean horizontalAsPercentage,
+                           Side verticalSide, double verticalPosition, boolean verticalAsPercentage) {
+        return this.horizontalSide == horizontalSide
+            && this.horizontalPosition == horizontalPosition
```

Comment:
so here `==` seems to be ok since we are comparing enums and primitive values, but...
should we account for floating point errors that might appear during interpolation?
i.e.
```
isClose(this.horizontalPosition, horizontalPosition) &&
```

where
```
private static boolean isClose(double a, double b) {
 return Math.abs(a - b) < EPSILON; // 0.00001 or something like that should be ok for screen coordinates
}
```

Reply:
I don't think that this is necessary. It would only save an allocation in a exceedingly rare case where the interpolation factor was microscopically different from either 0 or 1. This shouldn't happen in practice.